### PR TITLE
dns/dyndns: new provider 'dynv6.com'

### PR DESCRIPTION
**Pull request by rzerres**
this patchset introduces the DynDNS provider dynv6.com.
The service is free and does support ipv6 and ipv4 updates.
The API-Documentation can be found at

* https://dynv6.com/docs/apis

Updates are secured via a token, that is created via the WEB-GUI.

## Features
* Free for everyone
  Please be aware that dynv6 don't offer enterprise-grade DDoS
  protection and SLAs.
* support for ipv6 and ipv4 hosts
* Native IPv6 or 6to4
* support for SPF and TXT records with up to 512 bytes of data
* support for labels starting with underscore like _acme-challenge
* support for deleting of records

### CODE REVIEW

File: dns/dyndns/src/etc/inc/plugins.inc.d/dyndns.inc [hunk 1]
```diff
@@ -163,6 +165,7 @@ function dyndns_configure_client($conf)
         $dnsUpdateURL = "{$conf['updateurl']}",
         $forceUpdate = $conf['force'],
         $dnsZoneID = $conf['zoneid'],
+        /* $dnsToken = $conf['token'], */
```
Comment (fichtner) [inline dns/dyndns/src/etc/inc/plugins.inc.d/dyndns.inc]: Probably a comment from an earlier version of the patch? Please remove if unneeded

File: dns/dyndns/src/etc/inc/plugins.inc.d/dyndns/phpDynDNS.inc [hunk 2]
```diff
@@ -90,6 +91,8 @@
  *  regfish             - Last Tested: 15 August 2017
  *  regfish v6          - Last Tested: 15 August 2017
  *  Amazon Route53 v6   - Last Tested: 19 November 2017
+ *  dynv6               - Last Tested: 23 Juni 2019
```
Comment (fichtner) [inline dns/dyndns/src/etc/inc/plugins.inc.d/dyndns/phpDynDNS.inc]:
```suggestion
 *  dynv6               - Last Tested: 23 June 2019
```

File: dns/dyndns/src/etc/inc/plugins.inc.d/dyndns/phpDynDNS.inc [hunk 3]
```diff
@@ -90,6 +91,8 @@
...
+ *  dynv6 v6            - Last Tested: 23 Juni 2019
```
Comment (fichtner) [inline dns/dyndns/src/etc/inc/plugins.inc.d/dyndns/phpDynDNS.inc]:
```suggestion
 *  dynv6 v6            - Last Tested: 23 June 2019
```

File: dns/dyndns/src/etc/inc/plugins.inc.d/dyndns.inc [hunk 4]
```diff
@@ -110,6 +110,8 @@ function dyndns_list()
         'dyndns-custom' => 'DynDNS (custom)',
         'dyndns-static' => 'DynDNS (static)',
         'dyns' => 'DyNS',
+        'dynv6' => 'dynv6',
+        'dynv6' => 'dynv6 (v6)',
```
Comment (fichtner) [inline dns/dyndns/src/etc/inc/plugins.inc.d/dyndns.inc]:
```suggestion
        'dynv6-v6' => 'dynv6 (v6)',
```

File: dns/dyndns/src/etc/inc/plugins.inc.d/dyndns/phpDynDNS.inc [hunk 5]
```diff
@@ -216,10 +219,12 @@ class updatedns
                 }
                 break;
             case 'duckdns':
+	    case 'dnyv6':
+	    case 'dynv6-v6':
             case 'regfish':
             case 'regfish-v6':
                 if (!$dnsUser) {
-                    $this->_error(3);
+                    $this->_error(11);
```
Comment (fichtner) [inline dns/dyndns/src/etc/inc/plugins.inc.d/dyndns/phpDynDNS.inc]: Why is this necessary?

File: dns/dyndns/src/etc/inc/plugins.inc.d/dyndns/phpDynDNS.inc [hunk 6]
```diff
@@ -1302,6 +1334,9 @@ class updatedns
             case 10:
                 $error = "Dynamic DNS ({$this->_dnsHost}): No change in my IP address and/or " . $this->_dnsMaxCacheAgeDays . " days has not passed. Not updating dynamic DNS entry.";
                 break;
+            case 11:
+                $status = 'Dynamic DNS: (ERROR!) No Token Provided.';
```
Comment (fichtner) [inline dns/dyndns/src/etc/inc/plugins.inc.d/dyndns/phpDynDNS.inc]: The GUI is really restrictive in this regard.. the help label mentions "token" but ultimately the field label remains username so changing this can confuse users who don't see the help text.

File: dns/dyndns/src/www/services_dyndns_edit.php [hunk 7]
```diff
@@ -340,11 +341,12 @@ function is_dyndns_username($uname)
                     <td><a id="help_for_password" href="#" class="showhelp"><i class="fa fa-info-circle"></i></a> <?= gettext("Password") ?></td>
                     <td>
                       <input name="password" type="password" id="password" value="<?= $pconfig['password'] ?>" />
-                      <div class="hidden" data-for="help_for_password">
-                        <?=gettext('FreeDNS (freedns.afraid.org): Enter your "Authentication Token" provided by FreeDNS.') ?>
-                        <br /><?= gettext('Route 53: Enter your Secret Access Key.') ?>
-                        <br /><?= gettext('Duck DNS: Leave blank.') ?>
-                      </div>
+                        <div class="hidden" data-for="help_for_password">
```
Comment (fichtner) [inline dns/dyndns/src/www/services_dyndns_edit.php]: The indent here is off (it was ok before)

File: dns/dyndns/src/etc/inc/plugins.inc.d/dyndns/phpDynDNS.inc [hunk 8]
```diff
@@ -1158,6 +1176,20 @@ class updatedns
                         $this->_debug($data);
                 }
                 break;
+            case 'dynv6':
```
Comment (fichtner) [inline dns/dyndns/src/etc/inc/plugins.inc.d/dyndns/phpDynDNS.inc]: this would miss dynv6-v6 as well?

**Comment by fichtner** [comment]
@rzerres Thanks for your contribution! :)

I added some review comments and questions. The implementation itself looks good, but there is a bit of trouble with missing "dynv6-v6" usage and GUI concerns over "token" description use.

File: dns/dyndns/src/www/services_dyndns_edit.php (hunk 7 above)
Comment (rzerres) [reply dns/dyndns/src/www/services_dyndns_edit.php]: > The indent here is off (it was ok before)

Thanks, typo here. I did saw, that the indentation was "tab indented", as everything else is "char indented". So i did correct, but made the typo. Please recheck. (Not needed for the patch ....)

File: dns/dyndns/src/www/services_dyndns_edit.php [hunk 9]
```diff
@@ -342,8 +343,9 @@ function is_dyndns_username($uname)
                       <input name="password" type="password" id="password" value="<?= $pconfig['password'] ?>" />
                       <div class="hidden" data-for="help_for_password">
                         <?=gettext('FreeDNS (freedns.afraid.org): Enter your "Authentication Token" provided by FreeDNS.') ?>
-                        <br /><?= gettext('Route 53: Enter your Secret Access Key.') ?>
-                        <br /><?= gettext('Duck DNS: Leave blank.') ?>
+                          <br /><?= gettext('Route 53: Enter your Secret Access Key.') ?>
```
Comment (fichtner) [inline dns/dyndns/src/www/services_dyndns_edit.php]:
```suggestion
                        <br /><?= gettext('Route 53: Enter your Secret Access Key.') ?>
```

File: dns/dyndns/src/www/services_dyndns_edit.php [hunk 10]
```diff
@@ -342,8 +343,9 @@ function is_dyndns_username($uname)
...
+                          <br /><?= gettext('Duck DNS: Leave blank.') ?>
```
Comment (fichtner) [inline dns/dyndns/src/www/services_dyndns_edit.php]:
```suggestion
                        <br /><?= gettext('Duck DNS: Leave blank.') ?>
```

File: dns/dyndns/src/www/services_dyndns_edit.php [hunk 11]
```diff
@@ -342,8 +343,9 @@ function is_dyndns_username($uname)
...
+                          <br /><?= gettext('dynv6: Leave blank.') ?>
```
Comment (fichtner) [inline dns/dyndns/src/www/services_dyndns_edit.php]:
```suggestion
                        <br /><?= gettext('dynv6: Leave blank.') ?>
```

File: dns/dyndns/src/etc/inc/plugins.inc.d/dyndns/phpDynDNS.inc (hunk 5 above)
Comment (fichtner) [inline dns/dyndns/src/etc/inc/plugins.inc.d/dyndns/phpDynDNS.inc]:
```suggestion
                    $this->_error(3);
```

almost there :) let's dial this new error back. the intention is good, but I don't think it adds clarity. can also remove the new error message below

File: dns/dyndns/src/etc/inc/plugins.inc.d/dyndns/phpDynDNS.inc (hunk 8 above)
Comment (fichtner) [reply dns/dyndns/src/etc/inc/plugins.inc.d/dyndns/phpDynDNS.inc]: I believe this is still open :)

File: dns/dyndns/src/etc/inc/plugins.inc.d/dyndns/phpDynDNS.inc (hunk 5 above)
Comment (rzerres) [reply dns/dyndns/src/etc/inc/plugins.inc.d/dyndns/phpDynDNS.inc]: all right. So letd remove error handling back to original ....
will fixup with last commit.

File: dns/dyndns/src/etc/inc/plugins.inc.d/dyndns/phpDynDNS.inc (hunk 8 above)
Comment (rzerres) [reply dns/dyndns/src/etc/inc/plugins.inc.d/dyndns/phpDynDNS.inc]: yepp. fixed :)

**Comment by rzerres** [comment]
Hey Franco,
the implementation wasn't stable. So i took the time to investigate the ipv6 situation.
Once i figured out how to force usage of the correct WAN interface in combination with the needed IP-Address (ipv4 vs ipv6), i updated the routines. Changes are marginal, but curl options are significant.
While debugging, i wanted some more meaningful curl messages. Thus the little metric array. If this is to much bload, feel free to get rid of it. 

Tested, and now given the expected results.

cheers Ralf

**Comment by fichtner** [comment]
Merged manually as 470c63337198c, thanks a lot! ❤️
